lib/spack/spack/spec_index.py

```diff
@@ -0,0 +1,323 @@
+            hash_prefix, self.FULL_HASH_STRING_LENGTH,
+        )
+        self.hash_prefix = hash_prefix.lower()
+
+    def __repr__(self):
+        return '{0}(hash_prefix={1!r})'.format(type(self).__name__, self.hash_prefix)
+
+
+class CompleteHash(PartialHash):
+    @abstractproperty
+    def complete_hash(self):
+        # type: () -> str
+        pass
+
+
+class ConcreteHash(CompleteHash):
+    complete_hash = None  # type: str
+
+    def __init__(self, complete_hash):
+        # type: (str) -> None
+        assert len(complete_hash) == self.FULL_HASH_STRING_LENGTH, (
+            complete_hash, self.FULL_HASH_STRING_LENGTH,
+        )
+        self.complete_hash = complete_hash.lower()
+
+    @property
+    def hash_prefix(self):
+        # type: () -> str
+        return self.complete_hash
+
+    def as_abstract_spec(self):
+        # type: () -> spack.spec.Spec
+        return spack.spec.Spec('/{0}'.format(self.complete_hash))
+
+    def __repr__(self):
+        return '{0}(complete_hash={1!r})'.format(
+            type(self).__name__, self.complete_hash)
+
+
+class IndexQuery(object):
+    kwargs = None  # type: Dict[str, Any]
+
+    def __init__(self, **kwargs):
+        self.kwargs = kwargs
+
+    def __getattr__(self, name):
+        if name == 'kwargs':
+            return super(IndexQuery, self).__getattr__('kwargs')
+        return self.kwargs.get(name, None)
+
+    def __repr__(self):
+        return 'IndexQuery({0})'.format(', '.join(
+            '{0}={1}'.format(k, repr(v)) for k, v in self.kwargs.items()
+        ))
+
+    @classmethod
+    def from_query_args(cls, **kwargs):
+        # type: (Any) -> IndexQuery
+        for field, value in kwargs.items():
+            if value is any:
+                kwargs[field] = None
+        if 'query_spec' in kwargs:
+            kwargs['query_specs'] = kwargs.pop('query_spec')
+        return cls(**kwargs)
+
+    def to_query_args(self):
+        # type: () -> Dict
+        return dict(
+            query_spec=any if self.query_specs is None else self.query_specs,
+            known=any if self.known is None else self.known,
+            installed=True if self.installed is None else self.installed,
+            explicit=any if self.explicit is None else self.explicit,
+            start_date=self.start_date,
+            end_date=self.end_date,
+            hashes=self.hashes,
+            for_all_architectures=(False if self.for_all_architectures is None
+                                   else self.for_all_architectures),
+        )
+
+
+@add_metaclass(ABCMeta)
+class SpecIndexable(object):
+    """Interface to look up and traverse known specs from multiple sources."""
+
+    @abstractmethod
+    def spec_index_lookup(self, hash_prefix):
+        # type: (PartialHash) -> Iterator[IndexEntry]
+        pass
+
+    @abstractmethod
+    def spec_index_query(self, query):
+        # type: (IndexQuery) -> Iterator[ConcretizedSpec]
+        pass
+
+
+class SpecIndex(object):
```

Comment:
right now this is IMO overly complex -- it needs to be refactored to just take a list of DB's, and to look up hashes in those.  Where the hashes come from should be someone else's concern.

the local DB is a DB, the remote DB is a DB, etc. -- they're all the same.  I don't think we need separate methods or anything for each kind of DB.  I was expecting that this would just look up *hashes* and map them to specs, from several DBs.